### PR TITLE
Fix batched prefill OOM: size padded_batch to actual users, not max_batch_size

### PR DESCRIPTION
**Pull request by arobergeTT**
## Problem description

Related tt-metal ticket: https://github.com/tenstorrent/tt-metal/issues/39868

PR #35722 added batched prefill but always allocates tensors for all 32 batch slots, even when only 2 users need prefill. The input tensor `[32, seq_len]` is flattened to `[1, 1, 32*seq_len, hidden_dim]` in the model forward pass, and the `all_gather_async` call in `distributed_norm` tries to allocate an output buffer that blows past available DRAM (92–96% consumed by KV cache).

**OOM error from gemma-3-4b-it on N300 (server log):**
```
TT_FATAL: Out of Memory: Not enough space to allocate 5368709120 B DRAM buffer
across 8 banks, where each bank needs to store 671088640 B, but bank size is
786432000 B (allocated: 754974400 B, free: 31457600 B)
```

**OOM error from gemma-3-27b-it on T3K (server log):**
```
TT_FATAL: Out of Memory: Not enough space to allocate 11274289152 B DRAM buffer
across 12 banks, where each bank needs to store 939524096 B, but bank size is
1021821792 B (allocated: 908831392 B, free: 112990400 B)
```

Both match the formula `32 × padded_seq_len × hidden_dim × 2 bytes` exactly:
| Model | Calculation | Buffer Size |
|---|---|---|
| gemma-3-4b-it (N300) | 32 × 32768 × 2560 × 2 | **5,368,709,120 B** (5.00 GiB) |
| gemma-3-27b-it (T3K) | 32 × 32768 × 3584 × 2 | **11,274,289,152 B** (10.5 GiB) |

### Root cause: `padded_batch` is hardcoded to `max_batch_size`

**Before (on `main`)** - `prefill_forward_text()`:
```python
use_batched_prefill = (
    batch_size > 1
    and len(set(prefill_seq_lens)) == 1
    and prefill_seq_lens[0] * batch_size < MAX_BATCHED_PREFILL_SEQ_LEN
    and self.data_parallel == 1
    and not getattr(self.model_args[0], "disable_batched_prefill", False)
)

all_users = [0] if use_batched_prefill else empty_slots
padded_batch = self.model_args[0].max_batch_size  # ← always 32
```

When 2 users are scheduled with 32K-token prompts, `batch_size=2` passes the guard (`2 * 32768 = 65536 < 131072`), but `padded_batch` is set to 32. The model allocates `[32, 32768]` input tensors → flattened to `[1, 1, 32×32768, H]` → OOM in `all_gather_async`.

### Warmup was also broken

**Before (on `main`)** - `warmup_model_prefill()`:
```python
logger.warning("Batched prefill in TTT is not supported")

for batch_size in [1, 32]:
    if batch_size == 32:
        # TODO: Remove continue when batched prefill is supported
        continue
```

Only `batch_size=1` was ever warmed up. Batched prefill traces were never pre-compiled, causing trace cache misses at runtime.

## What's changed
All changes in `models/tt_transformers/tt/generator.py` (1 file, +36/−13 lines):

### 1. New module-level constant

```python
SUPPORTED_PREFILL_BATCH_SIZES = (1, 2, 4, 8, 16, 32)
```

Single source of truth used by both warmup and runtime to ensure trace-compatible batch sizes stay in sync.

### 2. Warmup - iterate over all supported sizes with a token-limit guard

```python
for batch_size in SUPPORTED_PREFILL_BATCH_SIZES:
    if batch_size > 1 and batch_size * supported_length >= MAX_BATCHED_PREFILL_SEQ_LEN:
        logger.info(
            f"Skipping batched prefill warmup for batch_size={batch_size}, "
            f"seq_len={supported_length}: exceeds token limit"
        )
        continue
```

All valid `(batch_size, seq_len)` combinations are now pre-compiled. Unsafe ones are skipped automatically.

### 3. Runtime - size `padded_batch` to actual users

**How `batch_size` is determined (the chain):**

1. **vLLM Scheduler** (`vllm/v1/core/sched/scheduler.py`) selects N new prefill requests that fit within `max_num_batched_tokens` (131,072). For example, two 31K-token RULER prompts (62K total < 131K budget) → N=2.

2. **TT Model Runner** (`vllm/v1/worker/tt_model_runner.py`) builds an input tensor from the scheduled requests:
   ```python
   input_tokens = input_batch.token_ids_cpu_tensor[:num_reqs, :max_prefill_tokens]
   # shape: [num_reqs, max_prefill_tokens] → e.g., [2, 31272]
   ```

3. **Generator** (`generator.py`) reads `batch_size` directly from the token tensor:
   ```python
   batch_size, batch_seq_len = tokens.shape  # batch_size = num_reqs from scheduler
   ```

So `batch_size` = however many same-step prefill requests the vLLM scheduler decided to co-schedule. We don't control this, but the scheduler does. What we control is how `padded_batch` is computed from it:

```python
if use_batched_prefill:
    padded_batch = next(
        (b for b in SUPPORTED_PREFILL_BATCH_SIZES if b >= batch_size),
        self.model_args[0].max_batch_size,
    )
    if padded_batch * prefill_seq_lens[0] >= MAX_BATCHED_PREFILL_SEQ_LEN:
        logger.info(
            f"Batched prefill disabled: {padded_batch} x {prefill_seq_lens[0]} = "
            f"{padded_batch * prefill_seq_lens[0]} tokens exceeds limit {MAX_BATCHED_PREFILL_SEQ_LEN}"
        )
        use_batched_prefill = False

if not use_batched_prefill:
    padded_batch = self.model_args[0].max_batch_size
```

`padded_batch` is rounded up to the nearest power-of-2 from `SUPPORTED_PREFILL_BATCH_SIZES` so that traces can be reused. For example: `batch_size=3` → `padded_batch=4`, `batch_size=2` → `padded_batch=2`.

For the OOM case (2 users, 32K tokens): `padded_batch` = **2** instead of 32. Buffer drops from **5 GiB → ~335 MB** (4b) and **10.5 GiB → ~670 MB** (27b).

### 4. Page table - use `padded_batch` dimension

```python
# Before:
padded_page_table = torch.ones(self.model_args[0].max_batch_size, ...)  # always [32, ...]

# After:
batch_dim = padded_batch_size if padded_batch_size is not None else self.model_args[0].max_batch_size
padded_page_table = torch.ones(batch_dim, ...)  # matches actual padded_batch
```

### gemma-3-27b-it on T3K - RULER evals with this fix:

Waiting results from the Models CI dispatch: https://github.com/tenstorrent/tt-shield/actions/runs/23203172938 

### gemma-3-4b-it on N300 - RULER evals with this fix: 

All 175/175 RULER requests completed. Zero OOM errors.

**Warmup log showing batch sizes pre-compiled:**
```
Warming up prefill for sequence length: 128 for batch size: 1
Warming up prefill for sequence length: 128 for batch size: 2
Warming up prefill for sequence length: 128 for batch size: 4
Warming up prefill for sequence length: 1024 for batch size: 1
Warming up prefill for sequence length: 1024 for batch size: 2
Warming up prefill for sequence length: 1024 for batch size: 4
Warming up prefill for sequence length: 2048 for batch size: 1
Warming up prefill for sequence length: 2048 for batch size: 2
Warming up prefill for sequence length: 2048 for batch size: 4
```

### Proof: batched prefill IS used at runtime and no longer OOMs

Batched prefill is identifiable in logs when "Prefill seq len" appears **without** a preceding "Prefilling User" message (the sequential path prints "Prefilling User X" before each prefill; the batched path does not).

**Batched vs sequential prefill counts:**

| | Old run (before fix) | New run (with fix) |
|---|---|---|
| Batched prefills | 36 | 16 |
| Sequential prefills | 141 (then OOM crash) | 1,095 (completed) |
| Outcome | `TT_FATAL` at 8K batched | All 175 RULER requests pass |

The old run crashed on the **first batched 8K prefill**. The new run processes batched prefills at all context lengths up to 32K without issue.

### The smoking gun: batched 32K prefill that would have OOMed

The new run's server log shows a batched 32K prefill at 21:45:36 - the **exact same request pattern** (two ~31K-token prompts, both padded to 32768) that caused the 5 GiB OOM in the old code:

```
# Previous 65K request completes:
21:45:35.832 | Finished prefill for all users up to 64817 tokens, Starting decode...

# Batched 32K prefill starts immediately - NO "Prefilling User" = batched path:
21:45:36.377 | Prefill seq len: 32768, max_prefill_chunk_size: 131072, trace: False

# Completes successfully 18.6 seconds later:
21:45:54.938 | Finished prefill for all users up to 31272 tokens, Starting decode...
```

With the old code (`padded_batch=32`), this would have allocated `32 × 32768 × 2560 × 2 = 5.37 GiB` → OOM. With the fix (`padded_batch=2`), it allocates `2 × 32768 × 2560 × 2 = 335 MB` → fits in DRAM.

### Timing confirms `padded_batch` is right-sized

Batched 4K prefills in the old run take **23–31 seconds** (processing 32 × 4096 = 131K tokens with `padded_batch=32`), while the same batched 4K prefills in the new run take **1.4–3.9 seconds** (processing 2 × 4096 = 8K tokens with `padded_batch=2`):

**Old run - batched 4K prefills (`padded_batch=32`):**
```
18:02:19.919 → 18:02:51.359  (~31s)   # 32 × 4096 = 131K tokens
18:02:51.389 → 18:03:15.335  (~24s)
18:03:45.606 → 18:04:09.279  (~24s)
```

**New run - batched 4K prefills (`padded_batch=2`):**
```
21:54:28.295 → 21:54:32.177  (~3.9s)  # 2 × 4096 = 8K tokens
21:54:56.002 → 21:54:57.442  (~1.4s)
21:55:00.743 → 21:55:02.173  (~1.4s)
```

**7–22x faster** due to right-sized batch dimension, and critically, small enough to fit in DRAM.

### How the old run crashed

The old run survived batched 1K, 2K, and 4K prefills (the `all_gather` buffer was ~1.3 GiB at 4K, tight but fit). It crashed on the **first batched 8K prefill**:

```
18:21:52.217 | Prefill seq len: 8192  ← batched, no "Prefilling User"

18:22:06.617 | TT_FATAL: Out of Memory: Not enough space to allocate 2684354560 B
             | DRAM buffer across 12 banks (free: 321261600 B)
```

The requested buffer = `32 × 8192 × 2560 × 2 × 2 (all_gather) = 2,684,354,560 B` - **exact match**.

### RULER eval results

```
| gemma-3-4b-it | n300 | ruler | PASS ✅ | 89.86 | 1.46x published (61.4) |
```

| Context Length | Score |
|---|---|
| 4096 | 89.86% |
| 8192 | 92.76% |
| 16384 | 80.86% |
| 32768 | 73.14% |
| 65536 | 61.62% |

## Related

- tenstorrent/tt-inference-server#2389

## Test Plan

- [x] `gemma-3-4b-it` on N300: RULER evals PASS (175/175, score 89.86), zero OOM
- [x] Pre-commit hooks pass (black, autoflake, isort)
- [x] Batched prefill confirmed active at runtime (16 batched prefills including 32K context)
- [x] Timing confirms right-sized `padded_batch` (1.4s vs 24s for batched 4K)
- [x] `gemma-3-27b-it` on T3K: RULER evals
- [x] Full CI

## Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:adam/gemma-3-batched-prefill)
- [x] New/Existing tests provide coverage for changes
  - Verified via RULER evals (long-context accuracy benchmark) which exercises the exact code path that was OOM-ing

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -462,17 +469,33 @@ def prefill_forward_text(
...
             and not getattr(self.model_args[0], "disable_batched_prefill", False)
         )
 
+        if use_batched_prefill:
+            padded_batch = next(
+                (b for b in SUPPORTED_PREFILL_BATCH_SIZES if b >= batch_size),
+                self.model_args[0].max_batch_size,
+            )
+            if padded_batch * prefill_seq_lens[0] >= MAX_BATCHED_PREFILL_SEQ_LEN:
+                logger.info(
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/tt_transformers/tt/generator.py:500]: @arobergeTT is this AI comment relevant?

**Review by mtairum — CHANGES_REQUESTED** [review]
Changes look ok.

CC: @gwangTT for extra pair of 👀 

I'm only blocking the merge because we need to see this passing on the Model demos pipelines at the very least.

- [ ] Single card: https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml
- [ ] T3K: https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml
- [ ] Galaxy: https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml

Some of these tests will fail. If the fail signature is the same as in main (just compare against any recent scheduled run) then all good from me. If it's not on a scheduled main, make sure it's not introduced by this commit.

Ping when you have those runs finished and analyzed and I'll approve the PR.

File: models/tt_transformers/tt/generator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arobergeTT) [reply models/tt_transformers/tt/generator.py:500]: Good catch from Copilot. It's not triggerable for Gemma-3 models as they use `max_batch_size=32` which is in `SUPPORTED_PREFILL_BATCH_SIZES`, so `padded_batch` never exceeds it. But it's a valid edge case for future models with non-power-of-2 `max_batch_size` (e.g., 12 → `next()` picks 16, exceeding the limit). Added a guard that disables batched prefill when `padded_batch` > `max_batch_size`, falling back to safe sequential prefill. I will add a fix for this.

[the remark this answers is not in the document]
Comment (arobergeTT) [reply models/tt_transformers/tt/generator.py:500]: Since non-power-of-2 max_batch_size doesn't really arise on TT devices (trace caching and hardware tiling both expect power-of-2) and all current models use `max_batch_size=32`, I added a guard that falls back to sequential prefill in that edge case. Just a one-line safety net without adding complexity for a scenario that shouldn't occur in practice.

**Comment by arobergeTT** [comment]
@mtairum Thank you for the review! Here are the Demos CI runs: 

- Single Card: https://github.com/tenstorrent/tt-metal/actions/runs/23750213976 -> All pass 
- T3K: https://github.com/tenstorrent/tt-metal/actions/runs/23750220552 
     - Failed 3 compared to 6 from the latest scheduled run: https://github.com/tenstorrent/tt-metal/actions/runs/23722851691/attempts/1. 
     - The 3 failed tests are identical - gpt oss, llama 70b, qwen2.5 
- Galaxy: https://github.com/tenstorrent/tt-metal/actions/runs/23751595126 
     - Passed 1 more test compared to latest scheduled run: https://github.com/tenstorrent/tt-metal/actions/runs/23722839378 
      - The 3 failed tests are identical - gpt oss, qwen3 32b, qwen3 32b long context 
      
      
None of these failed tests are due to my changes. 
cc: @ssinghalTT

**Review by mtairum — APPROVED** [review]
